### ISSUE TITLE
How to train models made out of models using add_loss?

### ISSUE THREAD

**Issue by Phibedy** [keras-team/keras#9960]
Hi,

I would like to train a VAE like this:
```
vae = Model(i, decoder(encoder_sampled(i)),name="vae")
vae.add_loss(...)
vae.compile(...,loss=None)
vae.fit(...)
```
It outputs:
```
UserWarning: Output "decoder" missing from loss dictionary.
We assume this was done on purpose, and we will not be expecting any data to be passed to "decoder" during training.
```
Afterwards it crashes because it receives the backprop. loss. How would I add the decoder to the loss dictionary?
Unluckily I haven't found much documentation about add_loss and how to add the decoder to the loss dictionary.

As a workaround I give the input images as target images and define a function (vae_loss_func) which I use in compile afterwards.
Full example:
``` python
def get_vae():
    latent_dim = 12
    intermediate_dim = 128
    epsilon_std = 1.0
    kernel = 3
    
    #encoder
    encoder_input = Input(shape=(256, 256,1,),name="en_input")
    l=Conv2D(32, kernel, padding='same',activation="relu",name="conv_01")(encoder_input)
    l=BatchNormalization(name="bn_01")(l)
    l=Conv2D(32, kernel, padding='same',activation="relu",name="conv_02")(l)
    l=BatchNormalization(name="bn_02")(l)
    l=MaxPooling2D(name="mp_01")(l)

    l=Conv2D(48, kernel, padding='same',activation="relu",name="conv_11")(l)
    l=BatchNormalization(name="bn_11")(l)
    l=Conv2D(48, kernel, padding='same',activation="relu",name="conv_12")(l)
    l=BatchNormalization(name="bn_12")(l)
    l=MaxPooling2D(name="mp_11")(l)
    
    l=Conv2D(64, kernel, padding='same',activation="relu",name="conv_21")(l)
    l=BatchNormalization(name="bn_21")(l)
    l=Conv2D(64, kernel, padding='same',activation="relu",name="conv_22")(l)
    l=BatchNormalization(name="bn_22")(l)
    l=MaxPooling2D(name="mp_21")(l)

    l=Flatten()(l)
    l=Dense(intermediate_dim,activation="relu",name="dense_31")(l)
    #from https://github.com/keras-team/keras/blob/master/examples/variational_autoencoder_deconv.py
    z_mean = Dense(latent_dim,name="z_mean")(l)
    z_log_var = Dense(latent_dim,name="z_log_var")(l)


    def sampling(args):
        z_mean, z_log_var = args
        epsilon = K.random_normal(shape=(K.shape(z_mean)[0], latent_dim),
                                  mean=0., stddev=epsilon_std)
        return z_mean + K.exp(z_log_var) * epsilon
    
    # note that "output_shape" isn't necessary with the TensorFlow backend
    # so you could write `Lambda(sampling)([z_mean, z_log_var])`
    latent_sampled = Lambda(sampling, output_shape=(latent_dim,),name="lambda")([z_mean, z_log_var])

    encoder = Model(encoder_input,z_mean,name="encoder")
    encoder_sampled = Model(encoder_input,latent_sampled,name="encoder_sampled")

    #decoder
    decoder_input = Input(shape=(latent_dim,),name="de_input")
    l=Dense(intermediate_dim,activation="relu",name="dense_41")(decoder_input)
    l=Dense(32*32*64,activation="relu",name="dense_42")(l)
    l=Reshape((32,32,64),name="r_41")(l)

    l=UpSampling2D(name="up_51")(l)
    l=Conv2DTranspose(64, kernel, padding='same',activation="relu",name="convt_51")(l)
    l=BatchNormalization(name="bn_51")(l)
    l=Conv2DTranspose(64, kernel, padding='same',activation="relu",name="convt_52")(l)
    l=BatchNormalization(name="bn_52")(l)
    
    l=UpSampling2D(name="us_61")(l)
    l=Conv2DTranspose(48, kernel, padding='same',activation="relu",name="convt_61")(l)
    l=BatchNormalization(name="bn_61")(l)
    l=Conv2DTranspose(48, kernel, padding='same',activation="relu",name="convt_62")(l)
    l=BatchNormalization(name="bn_62")(l)

    l=UpSampling2D(name="us_71")(l)
    l=Conv2DTranspose(32, kernel, padding='same',activation="relu",name="convt_71")(l)
    l=BatchNormalization(name="bn_71")(l)
    o=Conv2DTranspose(1, kernel, padding='same',activation="relu",name="convt_72")(l)

    decoder = Model(decoder_input,o,name="decoder")

    #build autoencoder
    var_auto_encoder = Model(encoder_input, decoder(encoder_sampled(encoder_input)),name="var_auto_encoder")
    auto_encoder = Model(encoder_input, decoder(encoder(encoder_input)),name="auto_encoder")
    
    # Compute VAE loss
    xent_loss = 256 * 256 * metrics.binary_crossentropy(
        K.flatten(encoder_input),
        K.flatten(o))
    kl_loss = - 0.5 * K.sum(1 + z_log_var - K.square(z_mean) - K.exp(z_log_var), axis=-1)
    vae_loss = K.mean(xent_loss + kl_loss)
    #var_auto_encoder.add_loss(vae_loss)
    #Causes
    #/home/ga97beh/.local/lib/python3.5/site-packages/ipykernel_launcher.py:8:
    #UserWarning: Output "decoder" missing from loss dictionary.
    #We assume this was done on purpose, and we will not be expecting any data to be passed to "decoder"
    #during training.
    
    #workaround
    def vae_loss_func(x, y_true):
        xent_loss = 256 * 256 * metrics.binary_crossentropy(
            K.flatten(x),
            K.flatten(y_true))
        kl_loss = - 0.5 * K.sum(1 + z_log_var - K.square(z_mean) - K.exp(z_log_var), axis=-1)
        return K.mean(xent_loss + kl_loss)
    
    return var_auto_encoder,auto_encoder,encoder,decoder,encoder_sampled,vae_loss_func
```

**Comment by asterisk37n** [comment]
Same issue here.

**Comment by XinDongol** [comment]
same issue

**Comment by johmicrot** [comment]
Same issue on TF 2.0

**Comment by DaielChom** [comment]
Same issue on TF 2.1

**Comment by i2507** [comment]
Same issue on TF 2.0
